### PR TITLE
feat(nextid): add tip feature

### PR DESCRIPTION
**Pull request by UncleBill**
## Description

Closes [MASK-730](https://mask.atlassian.net/browse/MASK-730)

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/web3-providers/src/helpers.ts [hunk 1]
```diff
@@ -10,3 +10,8 @@ export async function fetchJSON<T = unknown>(requestInfo: RequestInfo, requestIn
     const res = await globalThis.fetch(requestInfo, requestInit)
     return res.json()
 }
+
+const CORS_PROXY = 'https://cors.r2d2.to'
+export function courier(url: string) {
+    return [CORS_PROXY, url].join('?')
```
Comment (Jack-Works) [inline packages/web3-providers/src/helpers.ts]: urlcat?

File: packages/mask/src/plugins/NextID/components/tip/NFTSection/NFTList.tsx [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                        toggleItem(token.tokenId)
+                    }}>
+                    <NFTCardStyledAssetPlayer
+                        chainId={chainId}
+                        contractAddress={token.contractDetailed.address}
+                        url={token.info.mediaUrl}
+                        tokenId={token.tokenId}
+                    />
+                    <Checkbox
+                        onChange={noop}
```
Comment (guanbinrui) [inline packages/mask/src/plugins/NextID/components/tip/NFTSection/NFTList.tsx:88]: Is it a TODO feature?

File: packages/mask/src/plugins/NextID/components/tip/NFTSection/index.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}))
+
+interface Props extends HTMLProps<HTMLDivElement> {}
+
+export const NFTSection: FC<Props> = ({ className, ...rest }) => {
+    const { t } = useI18N()
+    const { erc721Contract, setErc721Contract, erc721TokenId, setErc721TokenId } = useTip()
+    const { classes } = useStyles()
+    const [balance, setBalance] = useState(0)
+    const onSearch = useCallback(() => {}, [])
```
Comment (guanbinrui) [inline packages/mask/src/plugins/NextID/components/tip/NFTSection/index.tsx]: Is it a TODO feature?

File: packages/mask/src/social-network-adaptor/twitter.com/injection/PostActions/selector.ts [hunk 4]
```diff
@@ -0,0 +1,7 @@
+import type { LiveSelector } from '@dimensiondev/holoflows-kit'
+import { querySelector } from '../../utils/selector'
+
+type E = HTMLElement
+export const selector = ''
+export const postActionsSelector: () => LiveSelector<E, true> = () =>
```
Comment (guanbinrui) [inline packages/mask/src/social-network-adaptor/twitter.com/injection/PostActions/selector.ts]: Plz move this to `../../utils/selector`.

File: packages/mask/src/web3/UI/ERC721ContractSelectPanel.tsx [hunk 5]
```diff
@@ -82,12 +82,14 @@ export function ERC721ContractSelectPanel(props: ERC721TokenSelectPanelProps) {
...
-    onBalanceChange(balance)
+    useEffect(() => {
+        onBalanceChange(balance)
+    }, [onBalanceChange, balance])
 
     const loading = (loadingFromChain || loadingBalanceFromNFTscan) && !balance
 
     // #region select contract
-    const [id] = useState(uuid())
+    const [id] = useState(uuid)
```
Comment (guanbinrui) [inline packages/mask/src/web3/UI/ERC721ContractSelectPanel.tsx:113]: Do you really mean it?

Comment (UncleBill) [reply packages/mask/src/web3/UI/ERC721ContractSelectPanel.tsx:113]: Yes, no need to create a new but won't be used uuid after the initialization. (`uuid` won't be executed and no needed to next time)

File: packages/mask/src/social-network-adaptor/twitter.com/injection/PostActions/selector.ts (hunk 4 above)
Comment (UncleBill) [reply packages/mask/src/social-network-adaptor/twitter.com/injection/PostActions/selector.ts]: This file is no longer needed. I originally wanted to put the selector alongside to the usage.

File: packages/mask/src/plugins/NextID/components/tip/NFTSection/NFTList.tsx (hunk 2 above)
Comment (UncleBill) [reply packages/mask/src/plugins/NextID/components/tip/NFTSection/NFTList.tsx:88]: No, it's a controlled component, should set both `onChange` and `checked` (or `defaultValue` only)

File: packages/mask/src/plugins/NextID/components/tip/NFTSection/index.tsx (hunk 3 above)
Comment (UncleBill) [reply packages/mask/src/plugins/NextID/components/tip/NFTSection/index.tsx]: Oops, it's missed. Already added.

File: packages/mask/src/plugins/NextID/SNSAdaptor/index.tsx [hunk 6]
```diff
@@ -16,6 +19,22 @@ const sns: Plugin.SNSAdaptor.Definition = {
             },
         },
     ],
+    GlobalInjection() {
+        if (!Flags.next_id_tip_enabled) return null
```
Comment (guanbinrui) [inline packages/mask/src/plugins/NextID/SNSAdaptor/index.tsx:23]: You can use the plugin infra API.

File: packages/mask/src/plugins/NextID/SNSAdaptor/index.tsx [hunk 7]
```diff
@@ -16,6 +19,22 @@ const sns: Plugin.SNSAdaptor.Definition = {
...
+        return (
+            <RootContext>
+                <TipTaskManager />
+            </RootContext>
+        )
+    },
+    PostActions() {
+        if (!Flags.next_id_tip_enabled) return null
```
Comment (guanbinrui) [inline packages/mask/src/plugins/NextID/SNSAdaptor/index.tsx:31]: You can use the `enableRequirement` API of plugin infra.

File: packages/mask/src/plugins/NextID/components/tip/TipTaskManager.tsx [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+export const TipTaskManager: FC = ({ children }) => {
+    const [tasks, setTasks] = useState<Task[]>(EMPTY_LIST)
+
+    const removeTask = useCallback((task: Task) => {
+        setTasks((list) => list.filter((t) => t !== task))
+    }, [])
+
+    useEffect(() => {
+        PluginNextIdMessages.tipTask.on((task) => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/NextID/components/tip/TipTaskManager.tsx]: To tear down the listener.

```suggestion
        return PluginNextIdMessages.tipTask.on((task) => {
```

File: packages/mask/src/plugins/NextID/contexts/TipContext.tsx [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    setRecipient: noop,
+    tipType: TipType.NFT,
+    setTipType: noop,
+    recipients: [],
+    setRecipients: noop,
+    token: null,
+    setToken: noop,
+    amount: '',
+    setAmount: noop,
+    erc721TokenId: null,
```
Comment (guanbinrui) [inline packages/mask/src/plugins/NextID/contexts/TipContext.tsx:54]: This plugin will support multi-chains. Better to avoid using chain-specific naming.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**